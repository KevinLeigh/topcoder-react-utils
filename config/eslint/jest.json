{
  "extends": "./default.json",
  "env": {
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "global-require": 0,
    "import/no-dynamic-require": 0
  }
}