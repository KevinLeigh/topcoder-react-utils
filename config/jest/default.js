module.exports = {
  collectCoverage: true,
  collectCoverageFrom: [
    'src/**/*.{js,jsx}',
  ],
  coverageDirectory: '__coverage__',
  moduleNameMapper: {
    '\\.(scss|css)$': 'identity-obj-proxy',
  },
  rootDir: '../..',
  testPathIgnorePatterns: [
    '/node_modules/',
  ],
  transformIgnorePatterns: [
    '/node_modules/(?!topcoder-react-utils)',
  ],
  setupTestFrameworkScriptFile: '<rootDir>/config/jest/setup.js',
};
