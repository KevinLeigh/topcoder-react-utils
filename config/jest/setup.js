/* NOTE: Don't rely on Babel in this module to avoid problems with usage in
 * a host package. */

require('raf/polyfill');
