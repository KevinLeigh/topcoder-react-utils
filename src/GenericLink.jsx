/**
 * The GenericLink wraps around React Router's Link or NavLink component, to
 * automatically replace them by regular <a> elements when:
 * - The target reference points to another domain;
 * - User opts to open the reference in a new tab;
 * - User explicitely opts to use <a>.
 */

/* global document, window */

import PT from 'prop-types';
import React from 'react';
import Url from 'url-parse';

export default function GenericLink(props) {
  const {
    activeClassName,
    activeStyle,
    children,
    className,
    enforceA,
    exact,
    isActive,
    location,
    onClick,
    onMouseDown,
    openNewTab,
    replace,
    routerLinkType,
    strict,
    to,
  } = props;
  /* Renders the link as <a> element if either opted explicitely, or the link
   * should open a new tab, or it is an anchor reference. */
  if (enforceA || openNewTab || to.startsWith('#')) {
    return (
      <a
        className={className}
        href={to}
        onClick={onClick}
        onMouseDown={onMouseDown}
        rel={openNewTab ? 'noopener noreferrer' : ''}
        target={openNewTab ? '_blank' : ''}
      >{children}</a>
    );
  }

  /* Otherwise we render the link as React Router's Link or NavLink element. */
  return React.createElement(
    routerLinkType, {
      activeClassName,
      activeStyle,
      className,
      exact,
      isActive,
      location,
      onClick: (e) => {
        /* If a custom onClick(..) handler was provided we execute it. */
        if (onClick) onClick(e);

        /* If the URL leads to a different domain we make transition using JS,
         * preventing event processing by React Router. */
        const url = new Url(to);
        if (url.origin !== document.location.origin) {
          document.location = to;
          e.preventDefault();

        /* Scroll to the top-left corner of the page, just in case the link
         * references the same page we are already in. */
        } else window.scroll(0, 0);
      },
      onMouseDown,
      replace,
      strict,
      to,
    },
  );
}

GenericLink.defaultProps = {
  activeClassName: null,
  activeStyle: null,
  children: null,
  className: null,
  enforceA: false,
  exact: false,
  isActive: null,
  location: null,
  onClick: null,
  onMouseDown: null,
  openNewTab: false,
  replace: false,
  strict: false,
  to: '',
};

GenericLink.propTypes = {
  activeClassName: PT.string,
  activeStyle: PT.string,
  children: PT.node,
  className: PT.string,
  enforceA: PT.bool,
  exact: PT.bool,
  isActive: PT.func,
  location: PT.shape(),
  onClick: PT.func,
  onMouseDown: PT.func,
  openNewTab: PT.bool,
  replace: PT.bool,
  routerLinkType: PT.element.isRequired,
  strict: PT.bool,
  to: PT.oneOfType([PT.object, PT.string]),
};
