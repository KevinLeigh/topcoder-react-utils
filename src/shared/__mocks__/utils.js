import * as mock from '../utils';
import isomorphy from '../utils/__mocks__/isomorphy';

mock.isomorphy = isomorphy;

export default mock;
