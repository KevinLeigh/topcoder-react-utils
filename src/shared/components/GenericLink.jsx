/**
 * The GenericLink wraps around React Router's Link or NavLink component, to
 * automatically replace them by regular <a> elements when:
 * - The target reference points to another domain;
 * - User opts to open the reference in a new tab;
 * - User explicitely opts to use <a>.
 */

/* global document, window */

import _ from 'lodash';
import PT from 'prop-types';
import React from 'react';
import Url from 'url-parse';

export default function GenericLink(props) {
  const {
    children,
    className,
    enforceA,
    onClick,
    onMouseDown,
    openNewTab,
    routerLinkType,
    to,
  } = props;
  /* Renders the link as <a> element if either opted explicitely, or the link
   * should open a new tab, or it is an anchor reference. */
  if (enforceA || openNewTab || to.startsWith('#')) {
    return (
      <a
        className={className}
        href={to}
        onClick={onClick}
        onMouseDown={onMouseDown}
        rel="noopener noreferrer"
        target={openNewTab ? '_blank' : ''}
      >{children}</a>
    );
  }

  const linkProps = _.omit(props, [
    'children',
    'enforceA',
    'openNewTab',
    'routerLinkType',
  ]);

  /* Otherwise we render the link as React Router's Link or NavLink element. */
  return React.createElement(
    routerLinkType, {
      ...linkProps,
      onClick: (e) => {
        /* If a custom onClick(..) handler was provided we execute it. */
        if (onClick) onClick(e);

        /* If the URL leads to a different domain we make transition using JS,
         * preventing event processing by React Router. */
        const url = new Url(to);
        if (url.origin !== document.location.origin) {
          document.location = to;
          e.preventDefault();

        /* Scroll to the top-left corner of the page, just in case the link
         * references the same page we are already in. */
        } else window.scroll(0, 0);
      },
    }, children,
  );
}

GenericLink.defaultProps = {
  children: null,
  className: null,
  enforceA: false,
  onClick: null,
  onMouseDown: null,
  openNewTab: false,
  replace: false,
  to: '',
};

GenericLink.propTypes = {
  children: PT.node,
  className: PT.string,
  enforceA: PT.bool,
  onClick: PT.func,
  onMouseDown: PT.func,
  openNewTab: PT.bool,
  routerLinkType: PT.func.isRequired,
  to: PT.oneOfType([PT.object, PT.string]),
};
