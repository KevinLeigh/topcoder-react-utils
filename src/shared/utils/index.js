import * as isomorphy from './isomorphy';

export default {
  isomorphy,
};
