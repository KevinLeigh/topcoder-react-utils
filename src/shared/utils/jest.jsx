/**
 * Utilities for Jest tests.
 */

/* NOTE: At the moment here is no intention to use this module for runtime
 * execution, only for tests; thus we depend on Jest environment and dev
 * dependencies. */

/* global expect */

import PT from 'prop-types';
import React from 'react';
import TU from 'react-dom/test-utils';

/* eslint-disable import/no-extraneous-dependencies */
import Renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
/* eslint-enable import/no-extraneous-dependencies */

/**
 * Just an alias for TU.findRenderedDOMComponentWithClass(..).
 * @param {Object} dom
 * @param {String} className
 * @return {Object}
 */
export function findInDomByClass(dom, className) {
  return TU.findRenderedDOMComponentWithClass(dom, className);
}

/**
 * Just an alias for TU.scryRenderedDOMComponentsWithClass(..).
 * @param {Object} dom
 * @param {Stirng} className
 * @return {Array}
 */
export function findInDomManyByClass(dom, className) {
  return TU.scryRenderedDOMComponentsWithClass(dom, className);
}

/**
 * Auxiliary wrapper around ReactJS Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the rendered tree.
 */
export function render(component) {
  return Renderer.create(component).toJSON();
}

/* The Wrapper is necessary for the "renderDom(..)" function, because
 * the "renderIntoDocument(..)" function from "react-dom/test-utils" works
 * only with state components, so we have to wrap our ReactJS components into
 * such Wrapper. */
class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    return this.props.children;
  }
}

Wrapper.propTypes = {
  children: PT.node.isRequired,
};

/**
 * Renders given component into DOM, using "react-dom/test-utils".
 * @param {Object} component ReactJS component to render.
 * @return {Object} Rendered DOM.
 */
export function renderDom(component) {
  return TU.renderIntoDocument(<Wrapper>{component}</Wrapper>);
}

/**
 * Auxiliary wrapper around ReactJS Shallow Test Renderer.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of the shallow component's render tree.
 */
export function shallowRender(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

/**
 * Renders provided ReactJS component with ReactJS Shallow Test Renderer,
 * creates/tests its snapshot, and returns the render result.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON representation of shallow render.
 */
export function shallowSnapshot(component) {
  const res = shallowRender(component);
  expect(res).toMatchSnapshot();
  return res;
}

/**
 * Renders provided ReactJS component with ReactJS Test Renderer,
 * creates/tests the snapshot, and returns the render.
 * @param {Object} component ReactJS component to render.
 * @return {Object} JSON render of the component.
 */
export function snapshot(component) {
  const res = render(component);
  expect(res).toMatchSnapshot();
  return res;
}

export const simulate = TU.Simulate;
