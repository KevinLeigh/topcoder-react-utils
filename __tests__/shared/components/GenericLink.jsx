/* global document, window */

import GenericLink from 'components/GenericLink';
import PT from 'prop-types';
import React from 'react';
import { findInDomByClass, renderDom, simulate, snapshot } from 'utils/jest';

function Link(props) {
  return (
    <button
      onClick={props.onClick}
      className={props.className}
    >
      {JSON.stringify(props)}
    </button>
  );
}

Link.defaultProps = {
  className: null,
};

Link.propTypes = {
  className: PT.string,
  onClick: PT.func.isRequired,
};

describe('Matches snapshots', () => {
  test('in a minimalistic example', () => {
    snapshot((
      <GenericLink
        routerLinkType={Link}
        to="SOME/TEST/URL"
      >
        LINK
      </GenericLink>
    ));
  });
  test('when rendered as <a> element', () => {
    snapshot((
      <GenericLink
        enforceA
        routerLinkType={Link}
        to="SOME/TEST/URL"
      >
        LINK
      </GenericLink>
    ));
    snapshot((
      <GenericLink
        openNewTab
        routerLinkType={Link}
        to="SOME/TEST/URL"
      >
        LINK
      </GenericLink>
    ));
  });
});

test('onClick(..) works when rendered as custom <Link>', () => {
  window.scroll = jest.fn();
  const clickHandler = jest.fn();
  let doc = renderDom((
    <GenericLink
      className="LINK"
      onClick={clickHandler}
      routerLinkType={Link}
      to="SOME/TEST/URL"
    >
      LINK
    </GenericLink>
  ));
  let link = findInDomByClass(doc, 'LINK');
  simulate.click(link);
  expect(clickHandler).toHaveBeenCalled();

  const domain = 'https://some.domain.com';
  document.location.assign = jest.fn();
  doc = renderDom((
    <GenericLink
      className="LINK"
      routerLinkType={Link}
      to={domain}
    >
      LINK
    </GenericLink>
  ));
  link = findInDomByClass(doc, 'LINK');
  simulate.click(link);

  expect(document.location.assign).toHaveBeenCalledWith(domain);
  expect(window.scroll).toHaveBeenCalledTimes(1);
});
