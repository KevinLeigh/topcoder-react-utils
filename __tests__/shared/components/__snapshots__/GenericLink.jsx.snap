// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshots in a minimalistic example 1`] = `
<button
  className={null}
  onClick={[Function]}
  type="button"
>
  {"to":"SOME/TEST/URL","className":null,"onMouseDown":null,"replace":false,"children":"LINK"}
</button>
`;

exports[`Matches snapshots when rendered as <a> element 1`] = `
<a
  className={null}
  href="SOME/TEST/URL"
  onClick={null}
  onMouseDown={null}
  rel="noopener noreferrer"
  target=""
>
  LINK
</a>
`;

exports[`Matches snapshots when rendered as <a> element 2`] = `
<a
  className={null}
  href="SOME/TEST/URL"
  onClick={null}
  onMouseDown={null}
  rel="noopener noreferrer"
  target="_blank"
>
  LINK
</a>
`;
